src/main/java/org/rumbledb/items/parsing/ItemParser.java

```diff
@@ -62,13 +63,45 @@
     private static final DataType vectorType = new VectorUDT();
     public static final DataType decimalType = new DecimalType(30, 15); // 30 and 15 are arbitrary
 
-    public static Item getItemFromObject(JsonIterator object, ExceptionMetadata metadata) {
+    public static Item getItemFromObject(JsonReader<Object> object, ExceptionMetadata metadata, boolean isTopLevel) {
         try {
-            if (object.whatIsNext().equals(ValueType.STRING)) {
-                return ItemFactory.getInstance().createStringItem(object.readString());
+            if (isTopLevel) {
+                object.getNextToken();
             }
-            if (object.whatIsNext().equals(ValueType.NUMBER)) {
-                String number = object.readNumberAsString();
+            byte token = object.last();
+            if (token == '"') {
+                String s = object.readString();
+                if (!isTopLevel) {
+                    object.getNextToken();
+                }
+                return ItemFactory.getInstance().createStringItem(s);
+            }
+            if (token == '+' || token == '-' || token == '.' || (token >= '0' && token <= '9')) {
+                StringBuilder sb = new StringBuilder();
+                while (
+                    token == '+'
+                        || token == '-'
+                        || token == '.'
+                        || (token >= '0' && token <= '9')
+                        || token == 'e'
+                        || token == 'E'
+                ) {
+                    sb.append((char) token);
+                    try {
+                        token = object.getNextToken();
+                    } catch (IOException e) {
+                        if (!isTopLevel) {
+                            RumbleException ex = new ParsingException(
+                                    "Parsing error! , or ] or } expected but found " + (char) token,
+                                    metadata
+                            );
+                            ex.initCause(e);
+                            throw ex;
+                        }
+                        // Here atm this will slow down if this is a top-level, lone number.
```

Comment:
Using currentIndex and Length as suggested above, the efficiency problem with top level integers you've mentioned as the first point can be addressed.

I am curious what other conditions can trigger this IOException block here. If the out of bounds problems was the only issue, I believe this try/catch block can be eliminated as well.

Reply:
This is a great finding, Can. Many thanks. I solves the issue and IOException is no longer needed.